### ISSUE TITLE
ge6a - DOS of IG mint/burn because of _deltaHedgePosition() revert

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-02-smilee-finance-judging#142]
ge6a

medium

# DOS of IG mint/burn because of _deltaHedgePosition() revert

## Summary

The function _deltaHedgePosition() is called during mint/burn, and under certain circumstances, it initiates token swaps within the Vault. The quantity of tokens to be exchanged depends on several parameters, one of which is the side token amount available in the Vault. A malicious user can send a large amount of side tokens to the Vault via ERC20 transfer, thereby significantly increasing the parameter in _deltaHedgePosition(). A sufficiently large sum will result in slippage during the swap exceeding the permissible limit, causing the mint/burn transaction to be reverted.

## Vulnerability Detail

The problem for the attacker arises when there are multiple LPs in the beginning of the epoch; later the sent amount will be distributed among them proportionally to the shares they hold. Therefore, for the attack to be successful, the malicious user must be the only LP at the beginning of the epoch (or the others must have a very small percentage share). Even if new LPs join after the epoch starts, they will not receive funds from the direct transfer, due to the order of operations during epoch rolling. Thus, the attacker has a mechanism for DOS on mint/burn when deemed necessary. Several example scenarios, which do not exhaust all possibilities, include:

1. A sudden spike in price, where traders can lock in profits at the expense of LPs. The malicious user executes DOS on mint/burn, preventing traders from locking in profits from profitable positions. By the end of the epoch, the price returns to its normal levels, and the malicious LP incurs no losses.
2. Expectations of price increases due to positive market news; the malicious user executes DOS to prevent traders to open potentially profitable positions.

The amounts needed  to trigger the slippage protection are not very large. A detailed description of the amounts required to exceed the permissible slippage for the tokens described in the README file can be found in one of my other reports titled "Dos through large deposit." I briefly want to explain why I believe the two reports are not duplicates. Both reports utilize the ability to trigger the check for exceeding the permissible slippage, but this is not the root cause because it is not a vulnerability. The presence of slippage protection is mandatory and cannot be removed. Therefore, both reports use the same feature of the protocol but have different root causes, compositions, and fixes.

<details>
<summary>POC</summary>
First should replace the content of testnet/TestnetSwapAdapter.sol with the source code below in order to simulate slippage.

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.15;

import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
import {IExchange} from "../interfaces/IExchange.sol";
import {IPriceOracle} from "../interfaces/IPriceOracle.sol";
import {ISwapAdapter} from "../interfaces/ISwapAdapter.sol";
import {AmountsMath} from "../lib/AmountsMath.sol";
import {SignedMath} from "../lib/SignedMath.sol";
import {TestnetToken} from "../testnet/TestnetToken.sol";

import {console} from "forge-std/console.sol";


contract TestnetSwapAdapter is IExchange, Ownable {
    using AmountsMath for uint256;

    // Variables to simulate swap slippage (DEX fees and slippage together)
    // exact swap slippage (maybe randomly set at test setup or randomly changed during test), has the priority over the range
    int256 internal _exactSlippage; // 18 decimals
    // range to control a random slippage (between min and max) during a swap (see echidna tests)
    int256 internal _minSlippage; // 18 decimals
    int256 internal _maxSlippage; // 18 decimals

    IPriceOracle internal _priceOracle;

    error PriceZero();
    error TransferFailed();
    error InvalidSlippage();
    error Slippage();

    constructor(address priceOracle) Ownable() {
        _priceOracle = IPriceOracle(priceOracle);
    }

    function setSlippage(int256 exact, int256 min, int256 max) external onlyOwner {
        if (min > max || SignedMath.abs(exact) > 1e18 || SignedMath.abs(exact) > 1e18 || SignedMath.abs(exact) > 1e18) {
            revert InvalidSlippage();
        }
        _exactSlippage = exact;
        _minSlippage = min;
        _maxSlippage = max;
    }

    function changePriceOracle(address oracle) external onlyOwner {
        _priceOracle = IPriceOracle(oracle);
    }

    /// @inheritdoc IExchange
    function getOutputAmount(address tokenIn, address tokenOut, uint256 amountIn) external view returns (uint) {
        return _getAmountOut(tokenIn, tokenOut, amountIn);
    }

    /// @inheritdoc IExchange
    function getInputAmount(address tokenIn, address tokenOut, uint256 amountOut) external view returns (uint) {
        return _getAmountIn(tokenIn, tokenOut, amountOut);
    }

    /// @inheritdoc IExchange
    function getInputAmountMax(address tokenIn, address tokenOut, uint256 amountOut) external view returns (uint) {
        uint256 amountIn = _getAmountIn(tokenIn, tokenOut, amountOut);
        uint256 amountInSlip = slipped(amountIn, true);
        return amountIn > amountInSlip ? amountIn : amountInSlip;
    }

    /// @inheritdoc ISwapAdapter
    function swapIn(address tokenIn, address tokenOut, uint256 amountIn) external returns (uint256 amountOut) {
        if (!IERC20Metadata(tokenIn).transferFrom(msg.sender, address(this), amountIn)) {
            revert TransferFailed();
        }
        amountOut = _getAmountOut(tokenIn, tokenOut, amountIn);

        amountOut = slipped(amountOut, false);

        (uint256 amountOutMin, uint256 amountOutMax) = _slippedValueOut(tokenIn, tokenOut, amountIn);
        if (amountOut < amountOutMin || amountOut > amountOutMax) {
            revert Slippage();
        }

        TestnetToken(tokenIn).burn(address(this), amountIn);
        TestnetToken(tokenOut).mint(msg.sender, amountOut);
    }

    /// @inheritdoc ISwapAdapter
    function swapOut(
        address tokenIn,
        address tokenOut,
        uint256 amountOut,
        uint256 preApprovedAmountIn
    ) external returns (uint256 amountIn) {
        amountIn = _getAmountIn(tokenIn, tokenOut, amountOut);

        amountIn = slipped(amountIn, true);

        if (amountIn > preApprovedAmountIn) {
            revert InsufficientInput();
        }

        (uint256 amountInMax, uint256 amountInMin) = _slippedValueIn(tokenIn, tokenOut, amountOut);

        if (amountIn < amountInMin || amountIn > amountInMax) {
            revert Slippage();
        }

        if (!IERC20Metadata(tokenIn).transferFrom(msg.sender, address(this), amountIn)) {
            revert TransferFailed();
        }

        TestnetToken(tokenIn).burn(address(this), amountIn);
        TestnetToken(tokenOut).mint(msg.sender, amountOut);
    }

    function _getAmountOut(address tokenIn, address tokenOut, uint amountIn) internal view returns (uint) {
        uint tokenOutPrice = _priceOracle.getPrice(tokenIn, tokenOut);
        amountIn = AmountsMath.wrapDecimals(amountIn, IERC20Metadata(tokenIn).decimals());
        return AmountsMath.unwrapDecimals(amountIn.wmul(tokenOutPrice), IERC20Metadata(tokenOut).decimals());
    }

    function _getAmountIn(address tokenIn, address tokenOut, uint amountOut) internal view returns (uint) {
        uint tokenInPrice = _priceOracle.getPrice(tokenOut, tokenIn);

        if (tokenInPrice == 0) {
            // Otherwise could mint output tokens for free (no input needed).
            // It would be correct but we don't want to contemplate the 0 price case.
            revert PriceZero();
        }

        amountOut = AmountsMath.wrapDecimals(amountOut, IERC20Metadata(tokenOut).decimals());
        return AmountsMath.unwrapDecimals(amountOut.wmul(tokenInPrice), IERC20Metadata(tokenIn).decimals());
    }

    /// @dev "random" number for (testing purposes) between `min` and `max`
    function random(int256 min, int256 max) public view returns (int256) {
        uint256 rnd = block.timestamp;
        uint256 range = SignedMath.abs(max - min); // always >= 0
        if (rnd > 0) {
            return min + int256(rnd % range);
        }
        return min;
    }

    /// @dev returns the given `amount` slipped by a value (simulation of DEX fees and slippage by a percentage)
    function slipped(uint256 amount, bool directionOut) public view returns (uint256) {
        int256 slipPerc = _exactSlippage;

        if(amount > 400000e18)
        {
            slipPerc = 0.21e18;
        }
        else return amount;


        int256 out;
        if (directionOut) {
            out = (int256(amount) * (1e18 + slipPerc)) / 1e18;
        } else {
            out = (int256(amount) * (1e18 - slipPerc)) / 1e18;
        }
        if (out < 0) {
            out = 0;
        }

        return SignedMath.abs(out);
    }

    function getSlippage(address tokenIn, address tokenOut) public view returns (uint256 slippage) {
        // Default baseline value:
        if (slippage == 0) {
            return 0.02e18;
        }
    }

    function _slippedValueOut(
        address tokenIn,
        address tokenOut,
        uint256 amountIn
    ) private view returns (uint256 amountOutMin, uint256 amountOutMax) {
        uint256 amountOut = _getAmountOut(tokenIn, tokenOut, amountIn);
        amountOutMin = (amountOut * (1e18 - getSlippage(tokenIn, tokenOut))) / 1e18;
        amountOutMax = (amountOut * (1e18 + getSlippage(tokenIn, tokenOut))) / 1e18;
    }

    function _slippedValueIn(
        address tokenIn,
        address tokenOut,
        uint256 amountOut
    ) private view returns (uint256 amountInMax, uint256 amountInMin) {
        uint256 amountIn = _getAmountIn(tokenIn, tokenOut, amountOut);
        amountInMax = (amountIn * (1e18 + getSlippage(tokenIn, tokenOut))) / 1e18;
        amountInMin = (amountIn * (1e18 - getSlippage(tokenIn, tokenOut))) / 1e18;
    }

}
```

After that should put the following function in Scenarios.t.sol:

```solidity
    function testDosWithTransfer() public
    {
        string memory scenarioName = "scenario_1";
        bool isFirstEpoch = true;

        console.log(string.concat("Executing scenario: ", scenarioName));
        string memory scenariosJSON = _getTestsFromJson(scenarioName);

        console.log("- Checking start epoch");
        StartEpoch memory startEpoch = _getStartEpochFromJson(scenariosJSON);
        _checkStartEpoch(startEpoch, isFirstEpoch);

        uint256 amount = 810e18;

        TokenUtils.provideApprovedTokens(_admin, _vault.sideToken(), _liquidityProvider, address(_vault), amount, vm);
        vm.startPrank(_liquidityProvider);
        IERC20(_dvp.sideToken()).transfer(address(_vault), amount);

        Trade[] memory trades = _getTradesFromJson(scenariosJSON);
        for (uint i = 0; i < trades.length; i++) {
            console.log("- Checking trade number", i + 1);
            _checkTrade(trades[i]); //this will revert due to slippage
        }
    }
```

</details>

## Impact

Temporary DOS resulting in profits for the malicious user and losses for traders.

## Code Snippet

https://github.com/sherlock-audit/2024-02-smilee-finance/blob/3241f1bf0c8e951a41dd2e51997f64ef3ec017bd/smilee-v2-contracts/src/IG.sol#L183-L224

## Tool used

Manual Review

## Recommendation

A possible solution is to record the available amount of side tokens at the end of each epoch in a variable and update it during the execution of deltaHedgePosition(). This way, tokens sent via direct transfer will not participate in the calculations.

**Comment by nevillehuang** [comment]
Escalate in place of LSW and sponsor

**Comment by sherlock-admin2** [comment]
> Escalate in place of LSW and sponsor

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by gstoyanovbg** [comment]
Disagree with the escalation. I don't see how an arbitrageur can steal part of the donation. The vault.deposit() function doesn't mint shares upon depositing liquidity, but only after the epoch is rolled. Let's consider the following example:

1) Epoch E0. User A makes a deposit at the end of the epoch. At this point, there are no other LPs, or if there are any, they have an insignificant share compared to the amount deposited by User A.
2) Epoch E0 is rolled, and at this point, shares are minted for the deposits made during the epoch. Epoch E1 begins.
3) If significant deposits are made after User A's deposit, the attack can be halted, and withdrawal of the deposit can be initiated. Otherwise, a donation is made at some point according to the description in the report.
4) User B decides to attempt to steal the donation by depositing a significant amount. However, at this point, it's impossible due to the way sharePrice is calculated in _beforeRollEpoch(). 

    ```solidity
    uint256 outstandingShares = totalSupply() - _state.withdrawals.heldShares;
    ...
    uint256 sharePrice = VaultLib.pricePerShare(lockedLiquidity, outstandingShares, _shareDecimals);
    epochPricePerShare[getEpoch().current] = sharePrice;
    ```

    As seen, lockedLiquidity is divided by totalSupply() - _state.withdrawals.heldShares. User B's shares haven't been minted yet, so they won't participate in determining the share price, and therefore, User B won't receive a share of User A's donation.

**Comment by panprog** [comment]
@gstoyanovbg 
I didn't mean deposit - indeed, if user is the only LP, no other user can do anything with the donation by depositting or withdrawing.

Arbitrageur can do this by providing JIT liquidity in the underlying DEX, forcing the deltahedge of entire donation to be swapped at the edge of the slippage price, which will benefit the arbitrageur at the expense of the donation. For example, user donates 100 ETH when ETH = 1000 USDC. Max slippage is 2%. Then arbitrageur adds 98000 USDC at the price of about 980. Then any tiny trade will delta hedge - sell 100 ETH at the price of 980 without the revert with arbitrageur receiving 100 ETH for 98000 USDC. He then withdraws uniswap liquidity and exchanges 100 ETH to 100000 USDC, a profit of +2000 USDC while at the same time mitigating the issue (no revert happening).

Another way to fix this is for admin to increase the max slippage limit.

However, I agree that such issue is valid even if it's fixable by different means. I just think it's low, because of very low probability (that single party will be LP), difficulty and partial loss of funds for the attacker and ease of mitigation.

**Comment by gstoyanovbg** [comment]
@panprog 

Got it and i agree that this is a risk for the attacker. There are two options:

1) To accept a limited risk of 2% of the donation in pursuit of greater profit. It is not clear whether the arbitrageur is aware of this possibility.

2) The risk can be reduced by opening new positions in order to use up the available liquidity. In order not to incur losses from them, the malicious user can open opposing positions in another similar protocol. This way, the arbitrageur will not be able to trigger a swap in the same block because they will not be able to open a position with mint, and in order to trigger a burn, someone must already have a position created. In the next block, the malicious user can see the increased liquidity and can defend themselves by stopping the attack and balancing the Vault or by making an additional donation.

I think setting too high a slippage is not a solution. It will indeed make this attack more expensive, but it will harm normal LPs under certain normal circumstances. Regarding the chance for a single LP in the first epoch - the attack is possible and could be profitable even if there are a few LPs with a small percent of the Vault compared to the attacker. This increase the probability.

In my opinion, this is borderline Medium, and the described risks can be classified as 'certain external conditions or specific states', which is part of the definition of Medium issue in Sherlock. However, I will respect Sherlock's decision.

**Comment by cvetanovv** [comment]
I disagree with the escalation. For me this issue is valid. The sponsor's comments are in the context of the report being a duplicate of #22, but that's a separate issue. And however unlikely the possibility of the attack is it is possible.

**Comment by Czar102** [comment]
@gstoyanovbg could you elaborate on how is this issue different than #127? Aside from the fact that the trigger is a donation here instead of a deposit.

**Comment by gstoyanovbg** [comment]
@Czar102 

127 - Demonstrates a Dos on the epoch rolling mechanism by accumulating a sufficiently large amount of deposits in a single epoch to trigger slippage protection when executing the adjustBalances() function, causing the transaction to revert. The root cause of this problem is the incorrect execution of a balancing swap at the end of the epoch without taking the necessary measures to ensure that this swap does not trigger slippage protection and cause a Dos on the protocol.

142 - There is flaw in the way the amount of tokens to be exchanged during [delta hedging](https://drive.google.com/file/d/1f0gCAV69CaH5GpqQtxK2wjScRNSEt5pr/view) is calculated causing DoS on the mint/burn functions of IG. A malicious LP can exploit this opportunity, for example, by preventing traders from closing their positions in order to maximize their own profit at the expense of traders.
In the deltaHedgePosition() function, the amount of base tokens is the available liquidity for trading for the corresponding epoch minus the used liquidity. 

```solidity
  ...
  Amount memory availableLiquidity = liquidity.available(strike);
  (, uint256 sideTokensAmount) = IVault(vault).balances();
  ...
```

The available trading liquidity is calculated at the beginning of each epoch and cannot change during the epoch. Therefore, deposits made after the start of the epoch do not participate in the calculations. However, the same function takes the amount of side tokens through the Vault.balances() function, which returns the actual amount of side tokens at the time of calling the function. It is incorrect to assume that from the beginning of an epoch until its end, the amount of side tokens can only change through swap operations initiated by the delta hedging mechanism. As the report shows, donations can change the amount of side tokens, leading to profits for malicious LPs. The root cause of the problem described in this report is that real amount of side tokens in the vault is improperly used in delta hedging calculations..

**Comment by Czar102** [comment]
@cvetanovv @gstoyanovbg @nevillehuang isn't this one of the slippage related issues listed as "Liquidity on the DEX pool decrease significantly [causing delta hedge or roll-epoch to revert due to max slippage]" in [the part of the docs](https://docs.google.com/spreadsheets/d/1Ff7oeqtM8CjKmcV9OP8Q7HcFsGFgBPTYST8MfGfRzIs)?

If yes, I'm planning to accept the escalation and consider this issue invalid/informational.

**Comment by gstoyanovbg** [comment]
@Czar102 In [this comment](https://github.com/sherlock-audit/2024-02-smilee-finance-judging/issues/127#issuecomment-2025029831) i explained why believe that both 127 and 142  are not known issues and don't fall into the category "Liquidity on the DEX pool decrease significantly [causing delta hedge or roll-epoch to revert due to max slippage]". Specifically for this report i want to note that there aren't any restrictions for the donation size and the slippage protection could be triggered anytime, so that a decreases of the liquidity on the DEX is not a requirement for this attack to be successful.

**Comment by nevillehuang** [comment]
This seems to be a known issue

> DEX SWAP: issues related to impacts of fees and slippage are known and mitigated

And possibly under the "Delta hedge manipulation leading to fee drain" and/or "MEV delta hedge swap" in docs [here](https://docs.google.com/spreadsheets/d/1Ff7oeqtM8CjKmcV9OP8Q7HcFsGFgBPTYST8MfGfRzIs/edit#gid=984409625)

**Comment by gstoyanovbg** [comment]
I disagree with the latest comment. I'm leaving the descriptions of the two known issues mentioned in the previous comment here. In my opinion, this is entirely sufficient to show that they have nothing in common with this report except for a few common words in the description.

> An IG trader could, before trading, calculate the protocol delta and frontrun / sandwitch the delta hedge


> In certain scenarios, the delta of the protocol may change much from a trade to another due to price movements.
    This happens in particular for: 
    1) highly volatile tokens
    2) tokens with limited liquidity, whose price is much impacted by Smilee swaps themselves
    In such scenario, a bad actor can buy and sell tiny position of IG OTM to cause swaps to rebalance delta.
    This:
    1) impacts the vault due to DEX slippage and fees
    2) can be used by the bad actor to profit from MEV or JITL on the underlying DEX pool

**Comment by Czar102** [comment]
I understand that the description may not be complete, but that doesn't imply that this type of issues hasn't been taken out of scope in the following line, which is an answer of out of scope issues:
> DEX SWAP: issues related to impacts of fees and slippage are known and mitigated

Planning to accept the escalation and invalidate the issue.

**Comment by gstoyanovbg** [comment]
@Czar102  You have the right to interpret every word as you see fit.

For me (and likely for most participants in this contest), it's perfectly clear that issues #127 and #142 aren't mitigated and are not even hinted at anywhere. Thus, the claim that they are known issues lacks any substantiation. I've elaborated on why the assumptions suggesting these problems are duplicates of known issues are, mildly stated, groundless.

> DEX SWAP: issues related to impacts of fees and slippage are known and **MITIGATED**

I leave it to the people reading this discussion to discover for themselves where problems #127 and #142 are mitigated. I wasn't able by myself. I don't even want to comment on the logic of invalidating a report due to a feature (slippage protection) that is used in a scenario without being related to the root cause. 

This is my last comment in this discussion. Feel free to resolve both escalations anytime.

**Comment by Czar102** [comment]
Let me put special pressure on the fact that the quoted paragraph is put in the section designed for out-of-scope issues. This is a reason of my interpretation as is.

**Comment by Czar102** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [nevillehuang](https://github.com/sherlock-audit/2024-02-smilee-finance-judging/issues/142/#issuecomment-2004410150): accepted
